ansible/cloud_providers/gcp_infrastructure_deployment.yml

```diff
@@ -0,0 +1,193 @@
+---
+- name: Step 001 Deploy Infrastructure
+  hosts: localhost
+  connection: local
+  gather_facts: false
+  become: false
+  tags:
+    - step001
+    - deploy_infrastructure
+  tasks:
+    - name: Ensure gcloud is installed
+      environment:
+        PATH: /usr/bin
+      command: which gcloud
+      register: gcloud_result
+
+    - name: Fail if gcloud not available
+      fail:
+        msg: you need Google Cloud SDK installed
+      when: gcloud_result is failed
+
+    - name: Get SSH public key
+      set_fact:
+        ssh_key: "~/.ssh/{{key_name}}.pem"
+        ssh_key_data: "{{lookup('file', '~/.ssh/{{key_name}}.pub')}}"
+
+    - name: Set the destination for the template
+      set_fact:
+        t_dest: "{{output_dir}}/{{ env_type }}.{{ guid }}.{{cloud_provider}}_cloud_template"
+      tags:
+        - gcp_infrastructure_deployment
+        - validate_gcp_template
+        - gen_gcp_template
+
+    - name: Generate GCP deployment manager template
+      template:
+        src: "../configs/{{ env_type }}/files/cloud_providers/{{cloud_provider}}_cloud_template.j2"
+        dest: "{{t_dest}}"
+      tags:
+        - gcp_infrastructure_deployment
+        - validate_gcp_template
+        - gen_gcp_template
+
+    # use command line 'gcloud' to validate template and deploy
+    - name: Login to Google Cloud SDK
+      command: >-
+        gcloud auth
+        activate-service-account "{{ gcp_account }}"
+        --key-file="{{ gcp_credentials_file }}"
+        --project="{{ gcp_project_id }}"
+      environment:
+        PATH: /usr/bin
+      tags:
+        - validate_gcp_template
+        - create_inventory
+        - must
```

Comment:
How does this gcloud auth exactly work ?
it seems stateful.

How would that work with concurrent deployments to different gcp accounts in ansible tower ?

Is there a way to use environment variables instead ?

To give an example, here is how we authenticate with OSP:

```yaml
- environment:
    OS_AUTH_URL: "{{ osp_auth_url }}"
    OS_USERNAME: "{{ osp_auth_username }}"
    OS_PASSWORD: "{{ osp_auth_password }}"
    OS_PROJECT_NAME: "{{ osp_project_name }}"
    OS_PROJECT_DOMAIN_ID: "{{ osp_auth_project_domain }}"
    OS_USER_DOMAIN_NAME: "{{ osp_auth_user_domain }}"
  ...
```

And most ansible modules understand those.
That makes the playbooks stateless.

Reply:
There is an environment variable, I haven't used it with service accounts so I will test it out and convert after.

Reply:
This has been completed, it now loads the authentication from environment variables and uses a config directory with guid in the name.